### ISSUE TITLE
segfaut when leaving statistics dialog

### ISSUE THREAD

**Issue by sevu** [wesnoth/wesnoth#1977]
Wesnoth segfaults when leaving the statistics dialog (alt+s) with the close button or escape, same happens for the _scroll to_ button.
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000556f4e7a987b in gui2::dialogs::statistics_dialog::on_primary_list_select (this=0x7fff8f4075d0, window=...) at src/gui/dialogs/statistics_dialog.cpp:252
252		for(const auto& i : *main_stat_table_[selected_row]) {
```
backtrace
```
#0  0x0000556f4e7a987b in gui2::dialogs::statistics_dialog::on_primary_list_select(gui2::window&) (this=0x7fff8f4075d0, window=...) at src/gui/dialogs/statistics_dialog.cpp:252
#1  0x0000556f4e7ab8d4 in std::__invoke_impl<void, void (gui2::dialogs::statistics_dialog::*&)(gui2::window&), gui2::dialogs::statistics_dialog*&, gui2::window&>(std::__invoke_memfun_deref, void (gui2::dialogs::statistics_dialog::*&)(gui2::window&), gui2::dialogs::statistics_dialog*&, gui2::window&) (__f=
    @0x556f74e3c9d0: (void (gui2::dialogs::statistics_dialog::*)(gui2::dialogs::statistics_dialog * const, gui2::window &)) 0x556f4e7a9730 <gui2::dialogs::statistics_dialog::on_primary_list_select(gui2::window&)>, __t=@0x556f74e3c9e8: 0x7fff8f4075d0, __args#0=...) at /usr/include/c++/7.2.0/bits/invoke.h:73
#2  0x0000556f4e7ab6fc in std::__invoke<void (gui2::dialogs::statistics_dialog::*&)(gui2::window&), gui2::dialogs::statistics_dialog*&, gui2::window&>(void (gui2::dialogs::statistics_dialog::*&)(gui2::window&), gui2::dialogs::statistics_dialog*&, gui2::window&) (__fn=
    @0x556f74e3c9d0: (void (gui2::dialogs::statistics_dialog::*)(gui2::dialogs::statistics_dialog * const, gui2::window &)) 0x556f4e7a9730 <gui2::dialogs::statistics_dialog::on_primary_list_select(gui2::window&)>, __args#0=@0x556f74e3c9e8: 0x7fff8f4075d0, __args#1=...) at /usr/include/c++/7.2.0/bits/invoke.h:95
#3  0x0000556f4e7ab364 in std::_Bind<void (gui2::dialogs::statistics_dialog::*(gui2::dialogs::statistics_dialog*, std::reference_wrapper<gui2::window>))(gui2::window&)>::__call<void, gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&, 0ul, 1ul>(std::tuple<gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&>&&, std::_Index_tuple<0ul, 1ul>) (this=0x556f74e3c9d0, __args=...) at /usr/include/c++/7.2.0/functional:467
#4  0x0000556f4e7aae8a in std::_Bind<void (gui2::dialogs::statistics_dialog::*(gui2::dialogs::statistics_dialog*, std::reference_wrapper<gui2::window>))(gui2::window&)>::operator()<gui2::widget&, gui2::event::ui_event, bool&, bool&, void*, void>(gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&) (this=0x556f74e3c9d0, __args#0=..., __args#1=@0x7fff8f40663c: gui2::event::NOTIFY_MODIFIED, __args#2=@0x7fff8f4066c6: false, __args#3=@0x7fff8f4066c7: false, __args#4=@0x7fff8f406620: 0x0) at /usr/include/c++/7.2.0/functional:551
#5  0x0000556f4e7aaa31 in std::_Function_handler<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*), std::_Bind<void (gui2::dialogs::statistics_dialog::*(gui2::dialogs::statistics_dialog*, std::reference_wrapper<gui2::window>))(gui2::window&)> >::_M_invoke(std::_Any_data const&, gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&) (__functor=..., __args#0=..., __args#1=@0x7fff8f40663c: gui2::event::NOTIFY_MODIFIED, __args#2=@0x7fff8f4066c6: false, __args#3=@0x7fff8f4066c7: false, __args#4=@0x7fff8f406620: 0x0)
    at /usr/include/c++/7.2.0/bits/std_function.h:316
#6  0x0000556f4e2686c5 in std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*)>::operator()(gui2::widget&, gui2::event::ui_event, bool&, bool&, void*) const (this=0x556f76747f60, __args#0=..., __args#1=gui2::event::NOTIFY_MODIFIED, __args#2=@0x7fff8f4066c6: false, __args#3=@0x7fff8f4066c7: false, __args#4=0x0)
    at /usr/include/c++/7.2.0/bits/std_function.h:706
#7  0x0000556f4e264b67 in gui2::event::implementation::fire_event<std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*)>, decltype(nullptr)>(gui2::event::ui_event, std::vector<std::pair<gui2::widget*, gui2::event::ui_event>, std::allocator<std::pair<gui2::widget*, gui2::event::ui_event> > >&, gui2::widget*, gui2::widget*, decltype(nullptr)&&) (event=gui2::event::NOTIFY_MODIFIED, event_chain=std::vector of length 0, capacity 0, dispatcher=0x556f5c6f0d10, w=0x556f5c6f0d10, params#0=<error reading variable>)
    at src/gui/core/event/dispatcher_private.hpp:433
#8  0x0000556f4e2604e6 in gui2::event::fire_event<std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*)>, decltype(nullptr)>(gui2::event::ui_event, gui2::event::dispatcher*, gui2::widget*, decltype(nullptr)&&) (event=gui2::event::NOTIFY_MODIFIED, d=0x556f5c6f0d20, w=0x556f5c6f0d10, params#0=<error reading variable>)
    at src/gui/core/event/dispatcher_private.hpp:505
#9  0x0000556f4e25e0bf in gui2::event::dispatcher::fire(gui2::event::ui_event, gui2::widget&, void*) (this=0x556f5c6f0d20, event=gui2::event::NOTIFY_MODIFIED, target=...)
    at src/gui/core/event/dispatcher.cpp:178
#10 0x0000556f4dd3a58e in gui2::listbox::list_item_clicked(gui2::widget&) (this=0x556f5c6f0d10, caller=...) at src/gui/widgets/listbox.cpp:295
#11 0x0000556f4dd480b6 in std::__invoke_impl<void, void (gui2::listbox::*&)(gui2::widget&), gui2::listbox*&, gui2::widget&>(std::__invoke_memfun_deref, void (gui2::listbox::*&)(gui2::widget&), gui2::listbox*&, gui2::widget&) (__f=
    @0x556f63dc0960: (void (gui2::listbox::*)(gui2::listbox * const, gui2::widget &)) 0x556f4dd3a408 <gui2::listbox::list_item_clicked(gui2::widget&)>, __t=@0x556f63dc0970: 0x556f5c6f0d10, __args#0=...) at /usr/include/c++/7.2.0/bits/invoke.h:73
#12 0x0000556f4dd46f76 in std::__invoke<void (gui2::listbox::*&)(gui2::widget&), gui2::listbox*&, gui2::widget&>(void (gui2::listbox::*&)(gui2::widget&), gui2::listbox*&, gui2::widget&) (__fn=
    @0x556f63dc0960: (void (gui2::listbox::*)(gui2::listbox * const, gui2::widget &)) 0x556f4dd3a408 <gui2::listbox::list_item_clicked(gui2::widget&)>, __args#0=@0x556f63dc0970: 0x556f5c6f0d10, __args#1=...) at /usr/include/c++/7.2.0/bits/invoke.h:95
#13 0x0000556f4dd45f6e in std::_Bind<void (gui2::listbox::*(gui2::listbox*, boost::arg<1>))(gui2::widget&)>::__call<void, gui2::widget&, 0ul, 1ul>(std::tuple<gui2::widget&>&&, std::_Index_tuple<0ul, 1ul>) (this=0x556f63dc0960, __args=...) at /usr/include/c++/7.2.0/functional:467
#14 0x0000556f4dd448ae in std::_Bind<void (gui2::listbox::*(gui2::listbox*, boost::arg<1>))(gui2::widget&)>::operator()<gui2::widget&, void>(gui2::widget&) (this=0x556f63dc0960, __args#0=...) at /usr/include/c++/7.2.0/functional:551
#15 0x0000556f4dd43046 in std::_Function_handler<void (gui2::widget&), std::_Bind<void (gui2::listbox::*(gui2::listbox*, boost::arg<1>))(gui2::widget&)> >::_M_invoke(std::_Any_data const&, gui2::widget&) (__functor=..., __args#0=...) at /usr/include/c++/7.2.0/bits/std_function.h:316
#16 0x0000556f4dcbd991 in std::function<void (gui2::widget&)>::operator()(gui2::widget&) const (this=0x556f63dc0150, __args#0=...) at /usr/include/c++/7.2.0/bits/std_function.h:706
#17 0x0000556f4dcc118e in std::__invoke_impl<void, std::function<void (gui2::widget&)>&, gui2::widget&>(std::__invoke_other, std::function<void (gui2::widget&)>&, gui2::widget&) (__f=..., __args#0=...) at /usr/include/c++/7.2.0/bits/invoke.h:60
#18 0x0000556f4dcc0df0 in std::__invoke<std::function<void (gui2::widget&)>&, gui2::widget&>(std::function<void (gui2::widget&)>&, gui2::widget&) (__fn=..., __args#0=...)
    at /usr/include/c++/7.2.0/bits/invoke.h:95
#19 0x0000556f4dcc08e9 in std::_Bind<std::function<void (gui2::widget&)> (boost::arg<1>)>::__call<void, gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&, 0ul>(std::tuple<gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&>&&, std::_Index_tuple<0ul>) (this=0x556f63dc0150, __args=...) at /usr/include/c++/7.2.0/functional:467
#20 0x0000556f4dcbfefc in std::_Bind<std::function<void (gui2::widget&)> (boost::arg<1>)>::operator()<gui2::widget&, gui2::event::ui_event, bool&, bool&, void*, void>(gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&) (this=0x556f63dc0150, __args#0=..., __args#1=@0x7fff8f406bfc: gui2::event::NOTIFY_MODIFIED, __args#2=@0x7fff8f406c86: false, __args#3=@0x7fff8f406c87: false, __args#4=@0x7fff8f406be0: 0x0) at /usr/include/c++/7.2.0/functional:551
#21 0x0000556f4dcbe827 in std::_Function_handler<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*), std::_Bind<std::function<void (gui2::widget&)> (boost::arg<1>)> >::_M_invoke(std::_Any_data const&, gui2::widget&, gui2::event::ui_event&&, bool&, bool&, void*&&) (__functor=..., __args#0=..., __args#1=@0x7fff8f406bfc: gui2::event::NOTIFY_MODIFIED, __args#2=@0x7fff8f406c86: false, __args#3=@0x7fff8f406c87: false, __args#4=@0x7fff8f406be0: 0x0) at /usr/include/c++/7.2.0/bits/std_function.h:316
#22 0x0000556f4e2686c5 in std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*)>::operator()(gui2::widget&, gui2::event::ui_event, bool&, bool&, void*) const (this=0x556f63dbfed0, __args#0=..., __args#1=gui2::event::NOTIFY_MODIFIED, __args#2=@0x7fff8f406c86: false, __args#3=@0x7fff8f406c87: false, __args#4=0x0)
    at /usr/include/c++/7.2.0/bits/std_function.h:706
#23 0x0000556f4e264b67 in gui2::event::implementation::fire_event<std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*)>, decltype(nullptr)>(gui2::event::ui_event, std::vector<std::pair<gui2::widget*, gui2::event::ui_event>, std::allocator<std::pair<gui2::widget*, gui2::event::ui_event> > >&, gui2::widget*, gui2::widget*, decltype(nullptr)&&) (event=gui2::event::NOTIFY_MODIFIED, event_chain=std::vector of length 0, capacity 0, dispatcher=0x556f76673800, w=0x556f76673800, params#0=<error reading variable>)
    at src/gui/core/event/dispatcher_private.hpp:433
#24 0x0000556f4e2604e6 in gui2::event::fire_event<std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&, void*)>, decltype(nullptr)>(gui2::event::ui_event, gui2::event::dispatcher*, gui2::widget*, decltype(nullptr)&&) (event=gui2::event::NOTIFY_MODIFIED, d=0x556f76673810, w=0x556f76673800, params#0=<error reading variable>)
    at src/gui/core/event/dispatcher_private.hpp:505
#25 0x0000556f4e25e0bf in gui2::event::dispatcher::fire(gui2::event::ui_event, gui2::widget&, void*) (this=0x556f76673810, event=gui2::event::NOTIFY_MODIFIED, target=...)
    at src/gui/core/event/dispatcher.cpp:178
#26 0x0000556f4e49e71c in gui2::toggle_panel::set_value(unsigned int) (this=0x556f76673800, selected=1) at src/gui/widgets/toggle_panel.cpp:176
#27 0x0000556f4e3b6b83 in gui2::policy::select_action::selection::select(gui2::grid&, bool) (this=0x556f75d593c0, grid=..., select=true) at src/gui/widgets/generator.cpp:1035
#28 0x0000556f4e40d5c2 in gui2::generator<gui2::policy::minimum_selection::one_item, gui2::policy::maximum_selection::one_item, gui2::policy::placement::vertical_list, gui2::policy::select_action::selection>::set_item_selected(unsigned int, bool) (this=0x556f75d593a0, index=0, selected=true) at src/gui/widgets/generator_private.hpp:1055
#29 0x0000556f4e3f7bce in gui2::generator<gui2::policy::minimum_selection::one_item, gui2::policy::maximum_selection::one_item, gui2::policy::placement::vertical_list, gui2::policy::select_action::selection>::do_select_item(unsigned int) (this=0x556f75d593a0, index=0) at src/gui/widgets/generator_private.hpp:920
#30 0x0000556f4e3b0750 in gui2::policy::minimum_selection::one_item::create_item(unsigned int) (this=0x556f75d593a0, index=0) at src/gui/widgets/generator.cpp:52
#31 0x0000556f4e3f71ca in gui2::generator<gui2::policy::minimum_selection::one_item, gui2::policy::maximum_selection::one_item, gui2::policy::placement::vertical_list, gui2::policy::select_action::selection>::create_item(int, std::shared_ptr<gui2::builder_grid const>, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, t_string, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, t_string> > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, t_string, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, t_string> > > > > > const&, std::function<void (gui2::widget&)> const&) (this=0x556f75d593a0, index=-1, list_builder=std::shared_ptr (count 3, weak 0) 0x556f50e991a0, item_data=std::map with 3 elements = {...}, callback=...) at src/gui/widgets/generator_private.hpp:747
#32 0x0000556f4dd393a7 in gui2::listbox::add_row(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, t_string, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, t_string> > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, t_string, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, t_string> > > > > > const&, int) (this=0x556f5c6f0d10, data=std::map with 3 elements = {...}, index=-1) at src/gui/widgets/listbox.cpp:79
#33 0x0000556f4e7a85f6 in gui2::dialogs::statistics_dialog::add_stat_row(gui2::window&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> > > const&, bool) (this=0x7fff8f4075d0, window=..., type="Recruits", value=std::map with 0 elements, has_cost=true) at src/gui/dialogs/statistics_dialog.cpp:119
#34 0x0000556f4e7a9108 in gui2::dialogs::statistics_dialog::update_lists(gui2::window&) (this=0x7fff8f4075d0, window=...) at src/gui/dialogs/statistics_dialog.cpp:184
#35 0x0000556f4e7a80aa in gui2::dialogs::statistics_dialog::pre_show(gui2::window&) (this=0x7fff8f4075d0, window=...) at src/gui/dialogs/statistics_dialog.cpp:95
#36 0x0000556f4dcbb307 in gui2::dialogs::modal_dialog::show(CVideo&, unsigned int) (this=0x7fff8f4075d0, video=..., auto_close_time=0) at src/gui/dialogs/modal_dialog.cpp:85
#37 0x0000556f4e713b04 in gui2::dialogs::statistics_dialog::display(team const&, CVideo&) (current_team=..., video=...) at src/gui/dialogs/statistics_dialog.hpp:37
#38 0x0000556f4e7007d8 in events::menu_handler::show_statistics(int) (this=0x7fff8f408538, side_num=1) at src/menu_events.cpp:157
#39 0x0000556f4e6fad00 in play_controller::hotkey_handler::show_statistics() (this=0x556f510f8e40) at src/hotkey/hotkey_handler.cpp:76
#40 0x0000556f4e9b9585 in hotkey::command_executor::execute_command(hotkey::hotkey_command const&, int, bool) (this=0x556f510f8e40, cmd=..., press=true)
    at src/hotkey/command_executor.cpp:208
#41 0x0000556f4e6fb918 in play_controller::hotkey_handler::execute_command(hotkey::hotkey_command const&, int, bool) (this=0x556f510f8e40, cmd=..., index=-1, press=true)
    at src/hotkey/hotkey_handler.cpp:263
#42 0x0000556f4e9babce in hotkey::execute_command(hotkey::hotkey_command const&, hotkey::command_executor*, int, bool) (command=..., executor=0x556f510f8e40, index=-1, press=true)
    at src/hotkey/command_executor.cpp:533
#43 0x0000556f4e9baab8 in hotkey::event_execute(SDL_Event const&, hotkey::command_executor*) (event=..., executor=0x556f510f8e40) at src/hotkey/command_executor.cpp:524
#44 0x0000556f4e9ba9af in hotkey::key_event(SDL_Event const&, hotkey::command_executor*) (event=..., executor=0x556f510f8e40) at src/hotkey/command_executor.cpp:508
#45 0x0000556f4e13c3c3 in controller_base::handle_event(SDL_Event const&) (this=0x7fff8f408310, event=...) at src/controller_base.cpp:62
#46 0x0000556f4e98e09e in events::pump() () at src/events.cpp:620
#47 0x0000556f4e13cfb9 in controller_base::play_slice(bool) (this=0x7fff8f408310, is_delay_enabled=true) at src/controller_base.cpp:239
#48 0x0000556f4e4fc0c7 in play_controller::play_slice_catch() (this=0x7fff8f408310) at src/play_controller.cpp:1061
#49 0x0000556f4e507bb2 in playsingle_controller::play_human_turn() (this=0x7fff8f408310) at src/playsingle_controller.cpp:472
#50 0x0000556f4e5071e9 in playsingle_controller::play_side_impl() (this=0x7fff8f408310) at src/playsingle_controller.cpp:390
#51 0x0000556f4e4fc828 in play_controller::play_side() (this=0x7fff8f408310) at src/play_controller.cpp:1154
#52 0x0000556f4e4fcad0 in play_controller::play_turn() (this=0x7fff8f408310) at src/play_controller.cpp:1192
#53 0x0000556f4e5057d3 in playsingle_controller::play_scenario_main_loop() (this=0x7fff8f408310) at src/playsingle_controller.cpp:184
#54 0x0000556f4e506127 in playsingle_controller::play_scenario(config const&) (this=0x7fff8f408310, level=...) at src/playsingle_controller.cpp:273
#55 0x0000556f4e239a11 in campaign_controller::playsingle_scenario(end_level_data&) (this=0x7fff8f408990, end_level=...) at src/game_initialization/playcampaign.cpp:194
#56 0x0000556f4e239f64 in campaign_controller::play_game() (this=0x7fff8f408990) at src/game_initialization/playcampaign.cpp:281
#57 0x0000556f4dc8f85b in game_launcher::launch_game(game_launcher::RELOAD_GAME_DATA) (this=0x556f506e05e0, reload=game_launcher::RELOAD_DATA) at src/game_launcher.cpp:949
#58 0x0000556f4dc015d7 in do_gameloop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (args=std::vector of length 2, capacity 2 = {...}) at src/wesnoth.cpp:832
#59 0x0000556f4dc01ed5 in main(int, char**) (argc=2, argv=0x7fff8f4098f8) at src/wesnoth.cpp:1065
```
build at acea15c312

**Comment by sevu** [comment]
Good news @Vultraz 
the cause lies as well in acea15c
#1975

**Comment by Vultraz** [comment]
GOD DAMMIT ALL TO HELL

**Comment by Vultraz** [comment]
Fixed: 085257f

**Comment by sevu** [comment]
Thanks
